File: ports/java/docproc/src/main/kotlin/net/russiandocs/docproc/modules/Deskewer.kt
package net.russiandocs.docproc.modules

import net.russiandocs.docproc.imaging.Contours
import net.russiandocs.docproc.imaging.Image
import net.russiandocs.docproc.imaging.Interpolation
import net.russiandocs.docproc.imaging.Io
import net.russiandocs.docproc.tensors.Ops
import org.opencv.core.Core
import org.opencv.core.CvType
import org.opencv.core.Mat
import org.opencv.core.Point as CvPoint
import org.opencv.core.Scalar
import org.opencv.core.Size
import org.opencv.imgproc.Imgproc
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * Removes residual tilt by scanning candidate angles and maximising the variance of the row-ink profile.
 *
 * A COARSE-TO-FINE search: a sparse scan across the whole range, then a dense scan around the coarse
 * winner. Verified in the reference to choose the same angles as a single dense scan while being about 2.9x
 * faster.
 */
public class DocDeskewer(
    private val angleRange: Double,
    angleSteps: Int,
    private val minAngle: Double,
    private val scale: Double,
    coarseSteps: Int,
) {
    private val coarseAngles: DoubleArray
    private val fineHalfRange: Double
    private val fineCount: Int

    init {
        val coarse = coarseSteps.coerceIn(3, angleSteps)
        coarseAngles = linspace(-angleRange, angleRange, coarse)

        val coarseStep = 2 * angleRange / (coarse - 1)
        val fullResolution = 2 * angleRange / (angleSteps - 1)
        fineHalfRange = coarseStep
        fineCount = max(3, (2 * coarseStep / fullResolution).roundToInt() + 1)
    }

    /**
     * Rotates the residual tilt out. Returns the corrected image and the angle applied.
     *
     * Below [minAngle] the image is returned unchanged — a copy, so ownership is uniform. The threshold is
     * load-bearing: it decides whether the canvas is rotated at all, and therefore whether every box
     * downstream lands in the same place.
     */
    public fun deskew(image: Image): Pair<Image, Double> {
        val angle = findAngle(image)
        if (abs(angle) < minAngle) {
            return image.clone() to angle
        }

        val rotation = rotationMatrix(image.width / 2.0, image.height / 2.0, angle, 1.0)
        try {
            val dst = Mat()
            Imgproc.warpAffine(image.mat, dst, rotation,
                Size(image.width.toDouble(), image.height.toDouble()),
                Imgproc.INTER_LINEAR, Core.BORDER_REPLICATE, Scalar(0.0))
            return Image.wrap(dst) to angle
        } finally {
            rotation.release()
        }
    }

    private fun findAngle(image: Image): Double {
        Io.toGray(image).use { gray ->
            val sh = max(1, (gray.height * scale).toInt())
            val sw = max(1, (gray.width * scale).toInt())
            Io.resize(gray, sw, sh, Interpolation.AREA).use { small ->
                val (binary, _) = Contours.thresholdOtsu(small, invert = true)
                binary.use {
                    val cx = sw / 2.0
                    val cy = sh / 2.0
                    val coarse = scoreAngles(binary, sw, sh, cx, cy, coarseAngles)
                    val ci = argmax(coarse)

                    // **A winner at either END of the coarse scan means no tilt.** The true optimum lies
                    // outside the search range, so refining around the edge would pick an arbitrary angle;
                    // the reference bails to 0.0 and this early exit is what keeps the canvas unrotated.
                    if (ci == 0 || ci == coarseAngles.size - 1) {
                        return 0.0
                    }

                    val best = coarseAngles[ci]
                    val lo = max(-angleRange, best - fineHalfRange)
                    val hi = min(angleRange, best + fineHalfRange)
                    val fineAngles = linspace(lo, hi, fineCount)
                    val fine = scoreAngles(binary, sw, sh, cx, cy, fineAngles)
                    return fineAngles[argmax(fine)]
                }
            }
        }
    }

    private fun scoreAngles(
        binary: Image,
        sw: Int,
        sh: Int,
        cx: Double,
        cy: Double,
        angles: DoubleArray,
    ): DoubleArray = DoubleArray(angles.size) { i ->
        val rotation = rotationMatrix(cx, cy, angles[i], 1.0)
        val rotated = Mat()
        try {
            // Nearest-neighbour and ZERO borders: the input is a binary mask, so interpolation would invent
            // grey values, and rotated-in area must contribute no ink.
            Imgproc.warpAffine(binary.mat, rotated, rotation, Size(sw.toDouble(), sh.toDouble()),
                Imgproc.INTER_NEAREST, Core.BORDER_CONSTANT, Scalar(0.0))
            // Row sums are exact integers, so this step contributes no float error at all — which makes the
            // variance the only numerically delicate part of the search. Ops.variance is TWO-PASS for
            // exactly that reason.
            Ops.variance(rowSumsOf(rotated))
        } finally {
            rotation.release()
            rotated.release()
        }
    }

    /**
     * Per-row ink totals of a single-channel 8-bit Mat, as exact integers.
     *
     * Read row-at-a-time rather than pixel-at-a-time: a JNI call per pixel is roughly 90 000 crossings on a
     * 300x300 scaled mask, times 21 coarse plus ~40 fine angles per document.
     */
    private fun rowSumsOf(binary: Mat): DoubleArray {
        val h = binary.rows()
        val w = binary.cols()
        val row = ByteArray(w)
        val sums = DoubleArray(h)
        for (y in 0 until h) {
            binary.get(y, 0, row)
            var acc = 0L
            for (x in 0 until w) {
                acc += (row[x].toInt() and 0xff).toLong()
            }
            sums[y] = acc.toDouble()
        }
        return sums
    }

    /** First maximum, like `np.argmax`. Strict `>` only. */
    private fun argmax(values: DoubleArray): Int {
        var best = 0
        var bestValue = Double.NEGATIVE_INFINITY
        for (i in values.indices) {
            if (values[i] > bestValue) {
                best = i
                bestValue = values[i]
            }
        }
        return best
    }

    private fun linspace(from: Double, to: Double, count: Int): DoubleArray {
        if (count <= 1) {
            return doubleArrayOf(from)
        }
        val step = (to - from) / (count - 1)
        return DoubleArray(count) { from + step * it }
    }

    /**
     * The rotation matrix.
     *
     * **Built by hand even though this port does not have to.** `Imgproc.getRotationMatrix2D` accepts a
     * double-valued `Point`, so D-08 does not apply here and the real function would work. It is written out
     * anyway so all four ports compute the identical matrix from the identical formula — the Go port has no
     * choice, and a difference here selects a different ANGLE, which rotates the canvas. Verified against
     * OpenCV's own output to 1.6e-14.
     */
    private fun rotationMatrix(cx: Double, cy: Double, angleDegrees: Double, scale: Double): Mat {
        val radians = angleDegrees * Math.PI / 180.0
        val alpha = Math.cos(radians) * scale
        val beta = Math.sin(radians) * scale

        val m = Mat(2, 3, CvType.CV_64FC1)
        m.put(0, 0, alpha)
        m.put(0, 1, beta)
        m.put(0, 2, (1 - alpha) * cx - beta * cy)
        m.put(1, 0, -beta)
        m.put(1, 1, alpha)
        m.put(1, 2, beta * cx + (1 - alpha) * cy)
        return m
    }

    public companion object {
        /**
         * The parameters the PIPELINE uses.
         *
         * Deliberately not the reference class's own defaults, which differ (its `angle_range` is 2.0).
         * Reading only the class would give a working deskewer that chooses different angles.
         */
        public fun forPipeline(): DocDeskewer = DocDeskewer(10.0, 101, 2.0, 0.4, 21)
    }
}

File: ports/dotnet/src/RussianDocs.DocumentProcessing/Modules/Deskewer.cs
using OpenCvSharp;
using RussianDocs.DocumentProcessing.Imaging;

namespace RussianDocs.DocumentProcessing.Modules;

/// <summary>
/// Removes residual tilt by scanning candidate angles and maximising the variance of the
/// row-ink profile.
///
/// <para>
/// A COARSE-TO-FINE search: a sparse scan across the whole range, then a dense scan around the coarse
/// winner. Verified in the reference to choose the same angles as a single dense scan while being
/// about 2.9x faster.
/// </para>
/// </summary>
public sealed class DocDeskewer
{
    private readonly double _angleRange;
    private readonly double _minAngle;
    private readonly double _scale;
    private readonly double[] _coarseAngles;
    private readonly double _fineHalfRange;
    private readonly int _fineCount;

    /// <summary>
    /// The parameters the PIPELINE uses.
    ///
    /// <para>
    /// Deliberately not the reference class's own defaults, which differ (its <c>angle_range</c> is
    /// 2.0). Reading only the class would give a working deskewer that chooses different angles.
    /// </para>
    /// </summary>
    public static DocDeskewer ForPipeline() => new(10.0, 101, 2.0, 0.4, 21);

    public DocDeskewer(double angleRange, int angleSteps, double minAngle, double scale,
        int coarseSteps)
    {
        _angleRange = angleRange;
        _minAngle = minAngle;
        _scale = scale;

        int coarse = Math.Clamp(coarseSteps, 3, angleSteps);
        _coarseAngles = Linspace(-angleRange, angleRange, coarse);

        double coarseStep = 2 * angleRange / (coarse - 1);
        double fullResolution = 2 * angleRange / (angleSteps - 1);
        _fineHalfRange = coarseStep;
        _fineCount = Math.Max(3, (int)Math.Round(2 * coarseStep / fullResolution) + 1);
    }

    /// <summary>
    /// Rotates the residual tilt out. Returns the corrected image and the angle applied.
    ///
    /// <para>
    /// Below <c>minAngle</c> the image is returned unchanged — a copy, so ownership is uniform. The
    /// threshold is load-bearing: it decides whether the canvas is rotated at all, and therefore
    /// whether every box downstream lands in the same place.
    /// </para>
    /// </summary>
    public (Image Deskewed, double Angle) Deskew(Image image)
    {
        double angle = FindAngle(image);
        if (Math.Abs(angle) < _minAngle)
        {
            return (image.Clone(), angle);
        }

        using Mat rotation = RotationMatrix(image.Width / 2.0, image.Height / 2.0, angle, 1.0);
        var dst = new Mat();
        Cv2.WarpAffine(image.Mat, dst, rotation, new Size(image.Width, image.Height),
            InterpolationFlags.Linear, BorderTypes.Replicate);
        return (Image.Wrap(dst), angle);
    }

    private double FindAngle(Image image)
    {
        using Image gray = Io.ToGray(image);

        int sh = Math.Max(1, (int)(gray.Height * _scale));
        int sw = Math.Max(1, (int)(gray.Width * _scale));
        using Image small = Io.Resize(gray, sw, sh, Interpolation.Area);
        (Image binary, double _) = Contours.ThresholdOtsu(small, invert: true);

        using (binary)
        {
            double cx = sw / 2.0, cy = sh / 2.0;
            double[] coarse = ScoreAngles(binary, sw, sh, cx, cy, _coarseAngles);
            int ci = Argmax(coarse);

            // **A winner at either END of the coarse scan means no tilt.** The true optimum lies
            // outside the search range, so refining around the edge would pick an arbitrary angle;
            // the reference bails to 0.0 and this early exit is what keeps the canvas unrotated.
            if (ci == 0 || ci == _coarseAngles.Length - 1)
            {
                return 0.0;
            }

            double best = _coarseAngles[ci];
            double lo = Math.Max(-_angleRange, best - _fineHalfRange);
            double hi = Math.Min(_angleRange, best + _fineHalfRange);
            double[] fineAngles = Linspace(lo, hi, _fineCount);
            double[] fine = ScoreAngles(binary, sw, sh, cx, cy, fineAngles);
            return fineAngles[Argmax(fine)];
        }
    }

    private static double[] ScoreAngles(Image binary, int sw, int sh, double cx, double cy,
        double[] angles)
    {
        var scores = new double[angles.Length];
        for (int i = 0; i < angles.Length; i++)
        {
            using Mat rotation = RotationMatrix(cx, cy, angles[i], 1.0);
            using var rotated = new Mat();
            // Nearest-neighbour and ZERO borders: the input is a binary mask, so interpolation would
            // invent grey values, and rotated-in area must contribute no ink.
            Cv2.WarpAffine(binary.Mat, rotated, rotation, new Size(sw, sh),
                InterpolationFlags.Nearest, BorderTypes.Constant, Scalar.All(0));
            scores[i] = Variance(RowSums(rotated));
        }
        return scores;
    }

    /// <summary>
    /// Per-row ink totals, as int64.
    ///
    /// <para>
    /// Exact integers, so this step contributes no float error at all — which is what makes the
    /// variance below the only numerically delicate part of the search.
    /// </para>
    /// </summary>
    private static long[] RowSums(Mat binary)
    {
        var sums = new long[binary.Rows];
        for (int y = 0; y < binary.Rows; y++)
        {
            long total = 0;
            for (int x = 0; x < binary.Cols; x++)
            {
                total += binary.Get<byte>(y, x);
            }
            sums[y] = total;
        }
        return sums;
    }

    /// <summary>
    /// numpy's TWO-PASS variance: subtract the mean, then average the squares.
    ///
    /// <para>
    /// **The one-pass form <c>E[x²] - E[x]²</c> is not equivalent here.** On values of order 255·W it
    /// loses about seven significant digits, which is enough to flip the argmax between two adjacent,
    /// nearly-equal angles — and the output is a discrete choice that rotates the image. Two passes,
    /// always.
    /// </para>
    /// </summary>
    private static double Variance(long[] values)
    {
        if (values.Length == 0)
        {
            return 0;
        }
        double mean = 0;
        foreach (long v in values)
        {
            mean += v;
        }
        mean /= values.Length;

        double acc = 0;
        foreach (long v in values)
        {
            double d = v - mean;
            acc += d * d;
        }
        return acc / values.Length;
    }

    /// <summary>First maximum, like <c>np.argmax</c>. Strict <c>&gt;</c> only.</summary>
    private static int Argmax(double[] values)
    {
        int best = 0;
        double bestValue = double.NegativeInfinity;
        for (int i = 0; i < values.Length; i++)
        {
            if (values[i] > bestValue)
            {
                best = i;
                bestValue = values[i];
            }
        }
        return best;
    }

    private static double[] Linspace(double from, double to, int count)
    {
        if (count <= 1)
        {
            return [from];
        }
        var values = new double[count];
        double step = (to - from) / (count - 1);
        for (int i = 0; i < count; i++)
        {
            values[i] = from + step * i;
        }
        return values;
    }

    /// <summary>
    /// The rotation matrix, built BY HAND rather than through <c>GetRotationMatrix2D</c>.
    ///
    /// <para>
    /// D-08. The reference passes a FRACTIONAL centre — <c>(sw/2.0, sh/2.0)</c> — and some bindings
    /// only accept an integer point, so the centre gets truncated. Measured in the Go spike: the
    /// integer version shifted the variance array by 3.8e-3 relative, which is ABOVE the 1e-3 policy,
    /// and therefore capable of selecting a different angle. This formula was verified against
    /// OpenCV's to 1.6e-14.
    /// </para>
    /// </summary>
    private static Mat RotationMatrix(double cx, double cy, double angleDegrees, double scale)
    {
        double radians = angleDegrees * Math.PI / 180.0;
        double alpha = Math.Cos(radians) * scale;
        double beta = Math.Sin(radians) * scale;

        var m = new Mat(2, 3, MatType.CV_64FC1);
        m.Set(0, 0, alpha);
        m.Set(0, 1, beta);
        m.Set(0, 2, (1 - alpha) * cx - beta * cy);
        m.Set(1, 0, -beta);
        m.Set(1, 1, alpha);
        m.Set(1, 2, beta * cx + (1 - alpha) * cy);
        return m;
    }
}
